src/AutoRest.CSharp/Common/Input/InputNamespaceVisitor.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using AutoRest.CSharp.Common.Input.Examples;
+using AutoRest.CSharp.Common.Utilities;
+
+namespace AutoRest.CSharp.Common.Input
+{
+    internal class InputNamespaceVisitor
+    {
+        protected virtual InputNamespace VisitNamespace(InputNamespace inputNamespace)
+        {
+            var typesCache = new Dictionary<InputType, InputType>();
+
+            var enums = VisitEnums(typesCache, inputNamespace.Enums);
+            var models = VisitModels(typesCache, inputNamespace.Models);
+            var clients = VisitClients(typesCache, inputNamespace.Clients);
+
+            return inputNamespace with
+            {
+                Clients = clients,
+                Models = models,
+                Enums = enums,
+            };
+        }
+
+        private IReadOnlyList<InputEnumType> VisitEnums(Dictionary<InputType, InputType> typesCache, IEnumerable<InputEnumType> source)
+        {
+            var targetEnums = new List<InputEnumType>();
+            foreach (var sourceEnum in source)
+            {
+                var targetEnum = VisitEnumType(sourceEnum);
+                targetEnums.Add(targetEnum);
+                typesCache[sourceEnum with {IsNullable = false}] = targetEnum with {IsNullable = false};
+            }
+            return targetEnums;
+        }
+
+        // Models have circular dependencies between them via BaseModel, DerivedModels, and CompositionModels properties
+        // Also, some of the InputModelProperty instances may reference InputModelType via Type property
+        // To correctly update all the references between models, this method does the following:
+        //   1. Sort models by their base models, so that models that have no base models are going first
+        //   2. Iterates over models in the said order and calls `VisitModel` for each of them. Resulted models still reference source models at that point, so all references are invalid
+        //   3. From the "intermediate" models collected on the previous step, create target models with correct `BaseModel` values and empty lists for DerivedModels, CompositionModels and Properties.
+        //      Target model is saved into `typesCache` and added to the returned list.
+        //      If `VisitModel` implementation has returned the same intermediate model for more than one source model, target model will be added only once
+        //   4. After all target models are created, lists for DerivedModels and CompositionModels are filled.
+        //      Also, each property is visited in `VisitModelProperty`, and if InputModelProperty.Type is a model, it must updated it to the target model from `typesCache`.
+        private IReadOnlyList<InputModelType> VisitModels(Dictionary<InputType, InputType> typesCache, IReadOnlyList<InputModelType> source)
```

Comment:
This seems overly complicated to avoid having a setter and instead restricting who can call that setter.  We should discuss our overall strategy here going forward.